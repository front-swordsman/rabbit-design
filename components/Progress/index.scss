@import '../styles/themes/default';

$progress: 'snake-progress';

.#{$progress} {
  width: 100%;
  position: relative;
  display: inline-block;

  &-line {
    .#{$progress} {
      &-container {
        padding-right: 60px;
        margin-right: -60px;
      }

      &-outer {
        border-radius: 16px;
        height: 8px;
      }

      &-inner {
        height: 8px;
        border-radius: 16px;
        top: 50%;
        margin-top: -4px;
      }
    }
  }

  &-small {
    .#{$progress} {
      &-outer {
        border-radius: 16px;
        height: 4px;
      }

      &-inner {
        height: 4px;
        border-radius: 16px;
        top: 50%;
        margin-top: -2px;
      }
    }

    &.#{$progress}-circle {
      width: 100px;
      height: 100px;
      font-size: 18px;

      &-outer {
        stroke-width: 6px;
      }

      &-inner {
        stroke-width: 6px;
      }
    }
  }

  &-large {
    .#{$progress} {
      &-outer {
        border-radius: 16px;
        height: 12px;
      }

      &-inner {
        height: 12px;
        border-radius: 16px;
        top: 50%;
        margin-top: -6px;
      }
    }

    &.#{$progress}-circle {
      width: 140px;
      height: 140px;
      font-size: 30px;

      &-outer {
        stroke-width: 10px;
      }

      &-inner {
        stroke-width: 10px;
      }
    }
  }

  &-circle {
    display: inline-block;
    position: relative;
    width: 120px;
    height: 120px;
    font-size: 24px;

    &-container {
      overflow: hidden;
    }

    &-outer {
      stroke-width: 8px;
      stroke: #ebebeb;
    }

    &-inner {
      transition: all 1s cubic-bezier(0.65, 0.2, 0.35, 1);
      stroke-linecap: round;
      stroke-width: 8px;
      stroke: $brand;
    }

    &-text {
      position: absolute;
      top: 50%;
      left: 50%;
      transform: translate(-50%, -50%);
      line-height: 1;
      color: #999;
      transition: transform 0.3s cubic-bezier(0.65, 0.2, 0.35, 1);
    }
  }

  &-container {
    display: inline-block;
    width: 100%;
    vertical-align: middle;
  }

  &-info {
    display: inline-block;
    vertical-align: middle;
    font-size: 14px;
    line-height: 1;
    color: #999;
    margin-left: 10px;
    word-wrap: normal;
  }

  &-outer {
    position: relative;
    width: 100%;
    background-color: #ebebeb;
  }

  &-inner {
    position: absolute;
    left: 0;
    top: 0;
    transition: all 1s cubic-bezier(0.65, 0.2, 0.35, 1);
    background: $brand;
  }

  &-success {
    .#{$progress}-inner {
      background-color: $successDeep;
    }
  }

  &-error {
    .#{$progress}-inner {
      background-color: $errorDeep;
    }
  }
}
