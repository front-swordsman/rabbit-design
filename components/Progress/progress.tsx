import * as React from 'react'
import cx from 'classnames'
import { ProgressProps } from 'types/progress'

import './index.scss'

const prefixCls = 'snake-progress'

const Progress: React.FC<ProgressProps> = ({
  size = 'medium',
  type = 'line',
  percent = 0,
  status = 'normal',
  showInfo = true,
  activeColor,
  width,
  textRender
}) => {
  const getClassNames = () => {
    return cx(prefixCls, {
      [`${prefixCls}-${type}`]: type,
      [`${prefixCls}-${size}`]: size,
      [`${prefixCls}-${status}`]: status,
      [`${prefixCls}-show-info`]: showInfo
    })
  }

  const getInnerStyle = () => {
    let style: React.CSSProperties = { width: `${percent}%` }
    if (activeColor) {
      style.backgroundColor = activeColor
    }
    return style
  }

  const getOuterStyle = () => {
    let style: React.CSSProperties = {}
    if (width) {
      style.width = width
    }
    return style
  }

  const renderCircleText = () => {
    if (textRender && typeof textRender === 'function') {
      return <div className={`${prefixCls}-circle-text`}>{textRender(percent)}</div>
    }
    return <div className={`${prefixCls}-circle-text`}>{`${percent}%`}</div>
  }

  const renderLine = () => {
    const innerStyle = getInnerStyle()
    const outStyle = getOuterStyle()
    return (
      <>
        <div className={`${prefixCls}-container`}>
          <div className={`${prefixCls}-outer`} style={outStyle}>
            <div className={`${prefixCls}-inner`} style={innerStyle} />
          </div>
        </div>
        {showInfo ? <div className={`${prefixCls}-info`}>{`${percent}%`}</div> : null}
      </>
    )
  }

  const renderCircle = () => {
    const style = activeColor ? { stroke: activeColor } : {}
    const strokeDashoffset = 289 * (1 - percent / 100)
    return (
      <>
        <svg className={`${prefixCls}-circle-container`} viewBox="0 0 100 100">
          <path
            className={`${prefixCls}-circle-outer`}
            d="M 50,50 m 0,-46 a 46,46 0 1 1 0,92 a 46,46 0 1 1 0,-92"
            fill-opacity="0"
          />
          <path
            className={`${prefixCls}-circle-inner`}
            d="M 50,50 m 0,-46 a 46,46 0 1 1 0,92 a 46,46 0 1 1 0,-92"
            fill-opacity="0"
            strokeDasharray="289 289"
            strokeDashoffset={strokeDashoffset}
            style={style}
          />
        </svg>
        {renderCircleText()}
      </>
    )
  }
  const classStr = getClassNames()
  return <div className={classStr}>{type === 'circle' ? renderCircle() : renderLine()}</div>
}

export default Progress
