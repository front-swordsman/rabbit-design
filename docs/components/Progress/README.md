---
imports: 
  import SimpleDemoCode from '!raw-loader!./simple.tsx';
  import SimpleDemo from './simple.tsx';
  import SizeDemoCode from '!raw-loader!./size.tsx';
  import SizeDemo from './size.tsx';
  import ColorDemoCode from '!raw-loader!./color.tsx';
  import ColorDemo from './color.tsx';
  import ShapeDemoCode from '!raw-loader!./shape.tsx';
  import ShapeDemo from './shape.tsx';
  import RenderDemoCode from '!raw-loader!./render.tsx';
  import RenderDemo from './render.tsx';
  import StatusDemoCode from '!raw-loader!./status.tsx';
  import StatusDemo from './status.tsx';
---

# Progress 进度条

上传文件，图片需要进度的时候使用

## 基础用法

:::demo
<Block code={SimpleDemoCode} des="基础用法">
<SimpleDemo />
</Block>
:::

## 不同尺寸

:::demo
<Block code={SizeDemoCode} des="基础用法">
<SizeDemo />
</Block>
:::

## 不同颜色

:::demo
<Block code={ColorDemoCode} des="基础用法">
<ColorDemo />
</Block>
:::

## 圆形进度条

:::demo
<Block code={ShapeDemoCode} des="基础用法">
<ShapeDemo />
</Block>
:::

## 自定义内部文本展示

:::demo
<Block code={RenderDemoCode} des="基础用法">
<RenderDemo />
</Block>
:::

## 不同的状态的进度条

:::demo
<Block code={StatusDemoCode} des="基础用法">
<StatusDemo />
</Block>
:::

# Api

| 参数        | 说明                 | 类型                           | 默认值   | 必选 |
| ----------- | -------------------- | ------------------------------ | -------- | ---- |
| size        | 进度条的尺寸         | `small`, `medium`, `large`     | `medium` | --   |
| type        | 进度条的形状         | `line`, `circle`               | `line`   | --   |
| percent     | 当前的进度           | number                         | `0`      | --   |
| showInfo    | 是否显示进度条的数值 | boolean                        | `true`   | --   |
| status      | 进度条的状态         | `normal`, `success`, `error`   | `normal` | --   |
| width       | 进度条的宽度         | number                         | --       | --   |
| activeColor | 进度条的颜色         | string                         | --       | --   |
| textRender  | 圆形进度条的内部文本 | (percent: number) => ReactNode | --       | --   |
