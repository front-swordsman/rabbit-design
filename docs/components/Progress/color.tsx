import * as React from 'react'
import { Progress, Button } from 'components'

const { useState } = React

export default function Simple() {
  const [percent, setPercent] = useState(0)

  const handleClick = () => {
    let p = percent
    p = percent >= 100 ? 0 : p + 10
    setPercent(p)
  }

  return (
    <div>
      <h2>不同颜色</h2>
      <div style={{ marginTop: 20 }}>
        <Progress percent={percent} activeColor="red" />
      </div>
      <div style={{ marginTop: 20 }}>
        <Progress percent={percent} activeColor="skyblue" />
      </div>
      <div style={{ marginTop: 20 }}>
        <Progress percent={percent} activeColor="pink" />
      </div>
      <Button onClick={handleClick}>点击增加进度</Button>
    </div>
  )
}
