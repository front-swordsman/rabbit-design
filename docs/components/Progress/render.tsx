import * as React from 'react'
import { Progress, Button } from 'components'

const { useState } = React

export default function Simple() {
  const [percent, setPercent] = useState(0)

  const handleClick = () => {
    let p = percent
    p = percent >= 100 ? 0 : p + 10
    setPercent(p)
  }

  return (
    <div>
      <Progress
        percent={percent}
        type="circle"
        textRender={percent => {
          if (percent === 100) return '成功'
          return 'loading'
        }}
      />
      <Button onClick={handleClick}>点击增加进度</Button>
    </div>
  )
}
