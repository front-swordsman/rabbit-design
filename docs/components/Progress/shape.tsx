import * as React from 'react'
import { Progress, Button } from 'components'

const { useState } = React

export default function Simple() {
  const [percent, setPercent] = useState(0)

  const handleClick = () => {
    let p = percent
    p = percent >= 100 ? 0 : p + 10
    setPercent(p)
  }

  return (
    <div>
      <h2>圆形</h2>
      <Progress percent={percent} type="circle" />
      <Progress percent={percent} size="small" type="circle" />
      <Progress percent={percent} size="large" type="circle" />
      <Button onClick={handleClick}>点击增加进度</Button>
    </div>
  )
}
