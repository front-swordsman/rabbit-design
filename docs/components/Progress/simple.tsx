import * as React from 'react'
import { Progress, Button } from 'components'

const { useState } = React

export default function Simple() {
  const [percent, setPercent] = useState(0)

  const handleClick = () => {
    let p = percent
    p = percent >= 100 ? 0 : p + 10
    setPercent(p)
  }

  return (
    <div>
      <Progress percent={percent} />
      <Button onClick={handleClick}>点击增加进度</Button>
    </div>
  )
}
