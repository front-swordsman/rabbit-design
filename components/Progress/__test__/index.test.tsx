import * as React from 'react'
import { render, fireEvent } from 'react-testing-library'
import Progress from '../index'

const { useState } = React

describe('Progress Test', () => {
  function Demo(props) {
    const [percent, setPercent] = useState(0)

    const handleClick = () => {
      let p = percent
      p = percent >= 100 ? 0 : p + 10
      setPercent(p)
    }

    return (
      <div>
        <Progress percent={percent} {...props} />
        <button onClick={handleClick}>点击增加进度</button>
      </div>
    )
  }
  it('snapshot with progress normal', () => {
    const { container } = render(<Progress />)
    expect(container.innerHTML).toMatchSnapshot()
  })

  it('snapshot with color progress', () => {
    const { container } = render(<Progress activeColor="red" />)
    expect(container.innerHTML).toMatchSnapshot()
  })

  it('snapshot with size large progress', () => {
    const { container } = render(<Progress size="large" />)
    expect(container.innerHTML).toMatchSnapshot()
  })

  it('snapshot with size small progress', () => {
    const { container } = render(<Progress size="small" />)
    expect(container.innerHTML).toMatchSnapshot()
  })

  it('snapshot with circle progress', () => {
    const { container } = render(<Progress type="circle" />)
    expect(container.innerHTML).toMatchSnapshot()
  })

  it('snapshot with success progress', () => {
    const { container } = render(<Progress status="success" />)
    expect(container.innerHTML).toMatchSnapshot()
  })

  it('snapshot with error progress', () => {
    const { container } = render(<Progress status="error" />)
    expect(container.innerHTML).toMatchSnapshot()
  })

  it('get correct progress', () => {
    const { getByText } = render(<Demo />)
    expect(getByText('0%')).not.toBeNull()
    const btn = getByText(/点击增加进度/i)
    fireEvent.click(btn)
    expect(getByText('10%')).not.toBeNull()
  })

  it('with text render', () => {
    const { getByText } = render(<Progress textRender={() => '成功'} type="circle" />)
    expect(getByText('成功')).not.toBeNull()
  })
})
