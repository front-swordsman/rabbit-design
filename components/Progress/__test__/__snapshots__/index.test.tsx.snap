// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Progress Test snapshot with circle progress 1`] = `"<div class=\\"snake-progress snake-progress-circle snake-progress-medium snake-progress-normal snake-progress-show-info\\"><svg class=\\"snake-progress-circle-container\\" viewBox=\\"0 0 100 100\\"><path class=\\"snake-progress-circle-outer\\" d=\\"M 50,50 m 0,-46 a 46,46 0 1 1 0,92 a 46,46 0 1 1 0,-92\\" fill-opacity=\\"0\\"></path><path class=\\"snake-progress-circle-inner\\" d=\\"M 50,50 m 0,-46 a 46,46 0 1 1 0,92 a 46,46 0 1 1 0,-92\\" fill-opacity=\\"0\\" stroke-dasharray=\\"289 289\\" stroke-dashoffset=\\"289\\"></path></svg><div class=\\"snake-progress-circle-text\\">0%</div></div>"`;

exports[`Progress Test snapshot with color progress 1`] = `"<div class=\\"snake-progress snake-progress-line snake-progress-medium snake-progress-normal snake-progress-show-info\\"><div class=\\"snake-progress-container\\"><div class=\\"snake-progress-outer\\"><div class=\\"snake-progress-inner\\" style=\\"width: 0%; background-color: red;\\"></div></div></div><div class=\\"snake-progress-info\\">0%</div></div>"`;

exports[`Progress Test snapshot with error progress 1`] = `"<div class=\\"snake-progress snake-progress-line snake-progress-medium snake-progress-error snake-progress-show-info\\"><div class=\\"snake-progress-container\\"><div class=\\"snake-progress-outer\\"><div class=\\"snake-progress-inner\\" style=\\"width: 0%;\\"></div></div></div><div class=\\"snake-progress-info\\">0%</div></div>"`;

exports[`Progress Test snapshot with progress normal 1`] = `"<div class=\\"snake-progress snake-progress-line snake-progress-medium snake-progress-normal snake-progress-show-info\\"><div class=\\"snake-progress-container\\"><div class=\\"snake-progress-outer\\"><div class=\\"snake-progress-inner\\" style=\\"width: 0%;\\"></div></div></div><div class=\\"snake-progress-info\\">0%</div></div>"`;

exports[`Progress Test snapshot with size large progress 1`] = `"<div class=\\"snake-progress snake-progress-line snake-progress-large snake-progress-normal snake-progress-show-info\\"><div class=\\"snake-progress-container\\"><div class=\\"snake-progress-outer\\"><div class=\\"snake-progress-inner\\" style=\\"width: 0%;\\"></div></div></div><div class=\\"snake-progress-info\\">0%</div></div>"`;

exports[`Progress Test snapshot with size small progress 1`] = `"<div class=\\"snake-progress snake-progress-line snake-progress-small snake-progress-normal snake-progress-show-info\\"><div class=\\"snake-progress-container\\"><div class=\\"snake-progress-outer\\"><div class=\\"snake-progress-inner\\" style=\\"width: 0%;\\"></div></div></div><div class=\\"snake-progress-info\\">0%</div></div>"`;

exports[`Progress Test snapshot with success progress 1`] = `"<div class=\\"snake-progress snake-progress-line snake-progress-medium snake-progress-success snake-progress-show-info\\"><div class=\\"snake-progress-container\\"><div class=\\"snake-progress-outer\\"><div class=\\"snake-progress-inner\\" style=\\"width: 0%;\\"></div></div></div><div class=\\"snake-progress-info\\">0%</div></div>"`;
